### PR TITLE
Sensor driver: PIR-based occupancy sensing

### PR DESCRIPTION
**Pull request by Hyungsin**
This PR provides a driver for EKMB PIR motion sensor.

The current driver operation is:
1) When the sensor detects motion, the connected GPIO pin goes high. Otherwise, it goes low.
2) The current driver supports to detect occupancy in % (how long has the GPIO been high?). 
3) It supports AUTO_INIT and SAUL.

### CODE REVIEW

**Comment by PeterKietzmann** [comment]
@Hyungsin pleas rebase. @MichelRottleuthner can you have a look at this PR? I've seen you playing with your PIR motion sensor so often, now please provide a review!

File: drivers/include/ekmb.h [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+/**
+ * @brief   Parameters needed for device initialization
+ */
+typedef struct {
+    gpio_t    gpio;            /**< GPIO pin that sensor is connected to */
```
Comment (MichelRottleuthner) [inline drivers/include/ekmb.h]: add a param for active-high/low

File: drivers/include/ekmb.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+} ekmb_params_t;
+
+/**
+  * @brief   Device descriptor for a EKMB device
+  * @{
+  */
+typedef struct {
+    ekmb_params_t p;
```
Comment (MichelRottleuthner) [inline drivers/include/ekmb.h]: move the global stete vars here

File: drivers/include/ekmb.h [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+} ekmb_t;
+/** @} */
+
+/**
+ * @brief   Initialize an EKMB device
+ *
+ * @param[out] dev          device descriptor
+ * @param[in] params        GPIO bus the device is connected to
+ *
+ * The valid address range is 0x1E - 0x1F depending on the configuration of the
```
Comment (MichelRottleuthner) [inline drivers/include/ekmb.h]: copy pasta?

File: drivers/ekmb/ekmb.c [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+int ekmb_init(ekmb_t *dev, const ekmb_params_t *params) {
+    dev->p.gpio = params->gpio;
+    pin_high = false;
+    accum_up_time = 0;
+    pin_rise_time = 0;
+    last_reset_time = xtimer_usec_from_ticks64(xtimer_now64());
+    if (gpio_init_int(params->gpio, GPIO_IN_PD, GPIO_BOTH, ekmb_trigger, dev)) {
+        return -1;
```
Comment (MichelRottleuthner) [inline drivers/ekmb/ekmb.c]: please use names for return codes

File: drivers/include/ekmb.h [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+ * address pins SA0 and SA1.
+ *
+ * @return                   0 on success
+ * @return                  -1 on error
+ * @return                  -2 on invalid address
```
Comment (MichelRottleuthner) [inline drivers/include/ekmb.h]: copy pasta?

File: drivers/include/ekmb.h [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+ */
+int ekmb_init(ekmb_t* dev, const ekmb_params_t* params);
+
+int ekmb_read(ekmb_t* dev, int16_t *light);
```
Comment (MichelRottleuthner) [inline drivers/include/ekmb.h]: *light?

**Comment by Hyungsin** [comment]
I first reflected your comments. Please check.
I think adding these features to the existing PIR can be a separate work.
The PIR driver seems like using an additional thread right? That may consume more memory, which I don't like.

**Comment by Hyungsin** [comment]
@MichelRottleuthner, I checked that the PIR driver does not necessarily add another thread, and revised this PR based on the existing PIR driver. I think it is good to be reviewed! :)

**Comment by tcschmidt** [comment]
@MichelRottleuthner changes o.k.?

File: drivers/pir/pir.c [hunk 7]
```diff
@@ -100,12 +131,30 @@ static void pir_callback(void *arg)
...
+        /* Add into accumulation */
+        dev->accum_active_time += (now - dev->start_active_time);
+    }
+    /* Pin is rising */
+    if (pin_now == dev->p.active_high) {
+        dev->start_active_time = now;
+        dev->active = true;
+    /* Pin is falling */
+    } else {
+        dev->start_active_time = 0xFFFFFFFFFFFFFFFF;
```
Comment (MichelRottleuthner) [inline drivers/pir/pir.c]: is this marking actuallly needed i.e. isn't the active flag sufficient? After all this callback is only called when the pin level changes, right?

File: sys/auto_init/saul/auto_init_pir.c [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+ */
+static pir_t pir_devs[PIR_NUM];
+
+/**
+ * @brief   Memory for the SAUL registry entries
+ */
+static saul_reg_t saul_entries[PIR_NUM];
+
+/**
+ * @brief   Reference the occupancy driver struct
```
Comment (MichelRottleuthner) [inline sys/auto_init/saul/auto_init_pir.c]: Reference **to**...

File: tests/driver_pir/main.c [hunk 9]
```diff
@@ -30,6 +30,7 @@
 
 static char pir_handler_stack[THREAD_STACKSIZE_MAIN];
 static pir_t dev;
+static const pir_params_t pir_param = {.gpio = PIR_GPIO, .active_high = true};
```
Comment (MichelRottleuthner) [inline tests/driver_pir/main.c]: don't define params here. Instead, include pir_params.h to use default parameters defined there. This also enables overwriting the active high parameter by CFLAGS commandline parameter. References to the old PIR_GPIO in the driver_pir's Makefile and main.c should then also be adapted to your newly introduced param defines (PIR_PARAM_GPIO and PIR_PARAM_ACTIVE_HIGH)

File: tests/driver_pir/main.c [hunk 10]
```diff
@@ -64,7 +65,7 @@ int main(void)
 {
     puts("PIR motion sensor test application\n");
     printf("Initializing PIR sensor at GPIO_%ld... ", (long)PIR_GPIO);
-    if (pir_init(&dev, PIR_GPIO) == 0) {
+    if (pir_init(&dev, &pir_param) == 0) {
```
Comment (MichelRottleuthner) [inline tests/driver_pir/main.c]: as stated above, use params from pir_params here (change to &pir_params[0])

File: drivers/include/pir.h [hunk 11]
```diff
@@ -28,12 +29,25 @@
...
 typedef struct {
-    gpio_t          gpio_dev;       /**< GPIO device which is used */
-    kernel_pid_t    msg_thread_pid; /**< thread to msg on irq */
+    kernel_pid_t msg_thread_pid; /**< thread to msg on irq */
+    bool active;                 /**< Indicate PIR is active or not */
+    uint64_t start_active_time;  /**< Time when PIR starts to be active */
+    uint64_t accum_active_time;  /**< Accumulated active time */
+    uint64_t last_read_time;     /**< Last time when PIR status is read */
+    pir_params_t p;              /**< Configuration parameters */
 } pir_t;
```
Comment (MichelRottleuthner) [inline drivers/include/pir.h]: I think reordering the struct members could save some memory here due to allignment/padding

File: drivers/include/pir.h [hunk 12]
```diff
@@ -62,12 +76,12 @@ typedef enum {
  * measurements can be made.
  *
  * @param[out] dev      device descriptor of an PIR sensor
- * @param[in] gpio      the GPIO device the sensor is connected to
+ * @param[in] params    parameters from the PIR sensor
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/include/pir.h]: done

File: drivers/include/pir.h [hunk 13]
```diff
@@ -78,6 +92,17 @@ int pir_init(pir_t *dev, gpio_t gpio);
  */
 pir_event_t pir_get_status(const pir_t *dev);
 
+/**
+ * @brief   Read OCCUPANCY value
+ *
+ * @param[in] dev       device descriptor of the PIR motion sensor to read from
+ * @param[out] occup    occupancy radio [in 100 * percentage]
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/include/pir.h]: done

File: drivers/pir/include/pir_params.h [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Set default configuration parameters for the PIR driver
+ * @{
+ */
+#ifndef PIR_PARAM_GPIO
+#define PIR_PARAM_GPIO             GPIO_PIN(PA, 6)
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/include/pir_params.h]: done

File: drivers/pir/include/pir_params.h [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+#endif
+#ifndef PIR_PARAM_ACTIVE_HIGH
+#define PIR_PARAM_ACTIVE_HIGH      1
+#endif
+
+#ifndef PIR_PARAMS
+#define PIR_PARAMS                 { .gpio        = PIR_PARAM_GPIO, \
+                                     .active_high = PIR_PARAM_ACTIVE_HIGH }
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/include/pir_params.h]: done

File: drivers/pir/pir.c [hunk 16]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
-    dev->gpio_dev = gpio;
+    dev->p.gpio = params->gpio;
+    dev->p.active_high = params->active_high;
     dev->msg_thread_pid = KERNEL_PID_UNDEF;
-    return gpio_init(dev->gpio_dev, GPIO_IN);
+
+    dev->active = false;
+    dev->accum_active_time = 0;
+    dev->start_active_time = 0;
+    dev->last_read_time = xtimer_usec_from_ticks64(xtimer_now64());
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: done

File: drivers/pir/pir.c [hunk 17]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
+    return 0;
 }
 
 pir_event_t pir_get_status(const pir_t *dev)
 {
-    return ((gpio_read(dev->gpio_dev) == 0) ? PIR_STATUS_LO : PIR_STATUS_HI);
+    return ((gpio_read(dev->p.gpio) == 0) ? PIR_STATUS_LO : PIR_STATUS_HI);
+}
+
+int pir_get_occupancy(const pir_t *dev, int16_t *occup) {
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: done

File: drivers/pir/pir.c [hunk 18]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
+    pir_t* pir_dev = (pir_t*) dev;
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: done

File: drivers/pir/pir.c [hunk 19]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
+    uint64_t now = xtimer_usec_from_ticks64(xtimer_now64());
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: done

File: drivers/pir/pir.c [hunk 20]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
+    uint64_t total_time = now - pir_dev->last_read_time;
+    if (total_time == 0) {
+        return -1;
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: done. changed all return values to some names.

File: sys/auto_init/saul/auto_init_pir.c (hunk 8 above)
Comment (Hyungsin) [reply sys/auto_init/saul/auto_init_pir.c]: done

File: tests/driver_pir/main.c (hunk 10 above)
Comment (Hyungsin) [reply tests/driver_pir/main.c]: done

File: tests/driver_pir/main.c (hunk 9 above)
Comment (Hyungsin) [reply tests/driver_pir/main.c]: done

File: drivers/pir/pir.c (hunk 7 above)
Comment (Hyungsin) [reply drivers/pir/pir.c]: I think so too. Thanks!

File: drivers/pir/pir.c [hunk 21]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
+    }
+
+    /* We were busy counting */
+    if (pir_dev->start_active_time != 0xFFFFFFFFFFFFFFFF && pir_dev->active) {
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: don't need it anymore

File: drivers/pir/pir.c [hunk 22]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
+    dev->p.active_high = params->active_high;
     dev->msg_thread_pid = KERNEL_PID_UNDEF;
-    return gpio_init(dev->gpio_dev, GPIO_IN);
+
+    dev->active = false;
+    dev->accum_active_time = 0;
+    dev->start_active_time = 0;
+    dev->last_read_time = xtimer_usec_from_ticks64(xtimer_now64());
+
+    if (gpio_init_int(dev->p.gpio, GPIO_IN_PD, GPIO_BOTH, pir_callback, dev)) {
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: I changed it configurable.

File: drivers/pir/pir.c [hunk 23]
```diff
@@ -37,16 +39,45 @@ static void pir_send_msg(pir_t *dev, pir_event_t event);
...
+        return -1;
+    }
+    return 0;
 }
 
 pir_event_t pir_get_status(const pir_t *dev)
 {
-    return ((gpio_read(dev->gpio_dev) == 0) ? PIR_STATUS_LO : PIR_STATUS_HI);
+    return ((gpio_read(dev->p.gpio) == 0) ? PIR_STATUS_LO : PIR_STATUS_HI);
```
[the remark this answers is not in the document]
Comment (Hyungsin) [reply drivers/pir/pir.c]: make sense. changed that.

File: drivers/include/pir.h (hunk 11 above)
Comment (Hyungsin) [reply drivers/include/pir.h]: What do you suggest? don't know padding that well.

**Comment by Hyungsin** [comment]
@MichelRottleuthner. Thank you!
I reflected most of your comments. Please go through it!

File: drivers/pir/pir.c [hunk 24]
```diff
@@ -48,57 +50,68 @@ int pir_init(pir_t *dev, const pir_params_t *params)
...
+    if (gpio_init_int(dev->p.gpio, gpio_mode, GPIO_BOTH, pir_callback, dev)) {
+        return PIR_NOGPIO;
+    }
+    return PIR_OK;
 }
 
 pir_event_t pir_get_status(const pir_t *dev)
 {
-    return ((gpio_read(dev->p.gpio) == 0) ? PIR_STATUS_LO : PIR_STATUS_HI);
+    return ((gpio_read(dev->p.gpio) == dev->p.active_high) ?
```
Comment (MichelRottleuthner) [inline drivers/pir/pir.c]: careful: the gpio API returns either 0 or >0. You could change `gpio_read(..)` to `(gpio_read(..) > 0)`

Comment (Hyungsin) [reply drivers/pir/pir.c]: done

**Comment by Hyungsin** [comment]
@MichelRottleuthner, resolved again. Thanks!

**Comment by MichelRottleuthner** [comment]
I just tested the github webeditor for resolving the remaining (minor) merge conflict. Seems like this was a bad idea and it messed something up :/ can you remove my commit and just do a proper rebase?

**Comment by Hyungsin** [comment]
No worries, done! :)

**Pull request merged**